src/target/ext/host_io.rs

```diff
@@ -0,0 +1,390 @@
+        None
+    }
+    /// Enable unlink operation.
+    #[inline(always)]
+    fn enable_unlink(&mut self) -> Option<HostIoUnlinkOps<Self>> {
+        None
+    }
+    /// Enable readlink operation.
+    #[inline(always)]
+    fn enable_readlink(&mut self) -> Option<HostIoReadlinkOps<Self>> {
+        None
+    }
+    /// Enable setfs operation.
+    #[inline(always)]
+    fn enable_setfs(&mut self) -> Option<HostIoSetfsOps<Self>> {
+        None
+    }
+}
+
+define_ext!(HostIoOps, HostIo);
+
+/// Nested Target Extension - Host I/O open operation.
+pub trait HostIoOpen: HostIo {
+    /// Open a file at filename and return a file descriptor for it, or return
+    /// [`HostIoError::Errno`] if an error occurs.
+    ///
+    /// The filename is a string, flags is an integer indicating a mask of open
+    /// flags (see [`HostIoOpenFlags`]), and mode is an integer indicating a
+    /// mask of mode bits to use if the file is created (see
+    /// [`HostIoOpenMode`]).
+    fn open(
+        &mut self,
+        filename: &[u8],
+        flags: HostIoOpenFlags,
+        mode: HostIoOpenMode,
+    ) -> HostIoResult<u32, Self>;
+}
+
+define_ext!(HostIoOpenOps, HostIoOpen);
+
+/// Nested Target Extension - Host I/O close operation.
+pub trait HostIoClose: HostIo {
+    /// Close the open file corresponding to fd.
+    fn close(&mut self, fd: u32) -> HostIoResult<(), Self>;
+}
+
+define_ext!(HostIoCloseOps, HostIoClose);
+
+/// Nested Target Extension - Host I/O pread operation.
+pub trait HostIoPread: HostIo {
+    /// Read data from the open file corresponding to fd.
+    ///
+    /// Up to count bytes will be read from the file, starting at offset
+    /// relative to the start of the file.
+    ///
+    /// The data read _must_ be sent by calling [`HostIoOutput::write`], which
+    /// will consume the `output` object and return a [`HostIoToken`]. This
+    /// token ensures that the implementer of this method calls
+    /// [`HostIoOutput::write`].
+    fn pread<'a>(
+        &mut self,
+        fd: u32,
+        count: <Self::Arch as Arch>::Usize,
+        offset: <Self::Arch as Arch>::Usize,
+        output: HostIoOutput<'a>,
+    ) -> HostIoResult<HostIoToken<'a>, Self>;
+}
+
+define_ext!(HostIoPreadOps, HostIoPread);
+
+/// Nested Target Extension - Host I/O pwrite operation.
+pub trait HostIoPwrite: HostIo {
+    /// Write data (a binary buffer) to the open file corresponding to fd.
+    ///
+    /// Start the write at offset from the start of the file.
+    ///
+    /// Unlike many write system calls, there is no separate count argument; the
+    /// length of data in the packet is used.
+    ///
+    /// Return the number of bytes written, which may be shorter
+    /// than the length of data, or [`HostIoError::Errno`] if an error occurred.
+    fn pwrite(
+        &mut self,
+        fd: u32,
+        offset: <Self::Arch as Arch>::Usize,
+        data: &[u8],
+    ) -> HostIoResult<u32, Self>;
```

Comment:
I considered about it, but don't we have `PacketSize` which is in limit size?

Reply:
Hmm, that's a interesting point, but I still think we should still switch to the more verbose signature. Even though the value is unlikely to ever exceed the size of a `u32`, we should still use the most "specific" type when possible.

It doesn't add all that much extra complexity to the user, especially since the recommended way to implement these method is to swap out instances of `<Self::Arch as Arch>::Usize` with the concrete type itself `e.g: u32` when implementing the method. It'll just make it marginally more complicated to parse the function signature, that's all.